src/tls13cert.rs

```diff
@@ -45,32 +45,51 @@ pub(crate) fn asn1_error<T>(err: Asn1Error) -> Result<T, Asn1Error> {
 // Long form length
 // * Must be used when the length is 128 or greater
 // * XXX: We do not accept lengths greater than 32-bit.
+#[hax_lib::ensures(|result| match result {
+    Result::Ok(_) => b.len() >= offset && b.len() - offset >= len,
+    _ => true })]
 fn long_length(b: &Bytes, offset: usize, len: usize) -> Result<usize, Asn1Error> {
     if len > 4 {
         asn1_error(ASN1_SEQUENCE_TOO_LONG)
-    } else {
+    } else if b.len() >= offset + len {
         let mut u32word = [U8(0); 4];
         u32word[0..len].copy_from_slice(&b[offset..offset + len]);
         Ok(u32_from_be_bytes(u32word).declassify() as usize >> ((4 - len) * 8))
+    } else {
+        asn1_error(ASN1_ERROR)
     }
 }
 
 // Read the length of a long form length
-fn length_length(b: &Bytes, offset: usize) -> usize {
-    if b[offset].declassify() >> 7 == 1u8 {
-        // Only in this case we have a length length.
-        (b[offset].declassify() & 0x7fu8) as usize
+#[hax_lib::ensures(|result| match result {
+    Result::Ok(l) => b.len() > offset && l <= 255,
+    _ => true })]
+fn length_length(b: &Bytes, offset: usize) -> Result<usize, Asn1Error> {
+    if b.len() > offset {
```

Comment:
Since we have early returns now, using that to error out early would make all this code much more readable. Doesn't have to happen in this PR though.